src/slicer.cpp

```diff
@@ -82,122 +84,305 @@ int SlicerLayer::getNextSegmentIdx(const Mesh* mesh, const SlicerSegment& segmen
+        int64_t dist2;
+        // polyline_idx*2 + {0 | front, 1 | back}
+        unsigned int terminus_0_idx;
+        unsigned int terminus_1_idx;
+
+        bool in_order() const
         {
-            PolygonRef open_polyline_other = open_polylines[open_polyline_other_idx];
-            
-            if (open_polyline_other.size() < 1) continue;
-            
-            Point diff = open_polyline.back() - open_polyline_other[0];
+            // in order if using back of line 1 and front of line 2
+            return (terminus_0_idx & 1) &&
+                !(terminus_1_idx & 1);
+        }
 
-            if (shorterThen(diff, largest_neglected_gap_second_phase))
+        // priority_queue will give greatest first so greatest
+        // must be most desirable stitch
+        bool operator<(const PossibleStitch &other) const
+        {
+            if (dist2 > other.dist2)
             {
-                if (open_polyline_idx == open_polyline_other_idx)
-                {
-                    polygons.add(open_polyline);
-                    open_polyline.clear();
-                    break;
-                }
-                else
-                {
-                    for (unsigned int line_idx = 0; line_idx < open_polyline_other.size(); line_idx++)
-                    {
-                        open_polyline.add(open_polyline_other[line_idx]);
-                    }
-                    open_polyline_other.clear();
-                }
+                return true;
+            }
+            else if (dist2 < other.dist2)
+            {
+                return false;
+            }
+
+            // use in order connections before reverse connections
+            if (!in_order() && other.in_order())
+            {
+                return true;
+            }
+
+            if (terminus_0_idx > other.terminus_0_idx)
+            {
+                return true;
+            }
+            else if (terminus_0_idx < other.terminus_0_idx)
+            {
+                return false;
+            }
+
+            if (terminus_1_idx > other.terminus_1_idx)
+            {
+                return true;
             }
+            else if (terminus_1_idx < other.terminus_1_idx)
+            {
+                return false;
+            }
+
+            return false;
         }
-    }
-}
+    };
+
+    std::priority_queue<PossibleStitch> stitch_queue;
 
-void SlicerLayer::stitch(Polygons& open_polylines)
-{ // TODO This is an inefficient implementation which can run in O(n^3) time.
-    // below code closes smallest gaps first
-    while(1)
     {
-        int64_t best_dist2 = max_stitch1 * max_stitch1;
-        unsigned int best_polyline_1_idx = -1;
-        unsigned int best_polyline_2_idx = -1;
-        bool reversed = false;
+        int64_t max_dist2 = max_dist * max_dist;
+
+        struct StitchGridVal
+        {
+            unsigned int polyline_idx;
+            Point polyline_end;
+        };
+
+        struct StitchGridValPointAccess
```

Comment:
It could have been done that way.  I feel like this way avoids repeating the array indexing multiple times at a trivial memory cost.